### ISSUE TITLE
Silencing warnings issued by numpy functions within dask.array

### ISSUE THREAD

**Issue by shoyer** [dask/dask#3245]
NumPy's NaN-skipping aggregation functions issue warnings when they encounter all NaN slices, e.g.,
```
In [2]: np.nanmean([np.nan])
/Users/shoyer/miniconda3/envs/xarray-py36/bin/ipython:1: RuntimeWarning: Mean of empty slice
  #!/Users/shoyer/miniconda3/envs/xarray-py36/bin/python
Out[2]: nan
```

As far as xarray is concerned, NaN is the right answer for these cases. So we silence these NumPy warnings, which works with NumPy, but unfortunately not with dask.array:
- Functions like `np.nanmean` issue warnings at compute time, outside our `warnings.catch_warnings()` context.
- Sometimes, dask arrays operations end up issuing new warnings (e.g., due to division by 0 when calculating the mean). 

Some possibilities for dealing with this:
1. Issue these irrelevant warnings, which needlessly annoys and distracts users. This is the current status quo.
2. Silence these warnings unilaterally, which could hide bugs in non-xarray code.
3. Rewrite the relevant dask.array functions to optionally call warning silenced version of NumPy functions. This would need to be in dask.array itself to be maintainable, but there isn't a good place in dask.array's API to inject lists of warnings to silence.
5. *Always* silence warnings when computing with dask, under the theory that warnings without good stacklevel identification are not so useful and  these get lost when computing with processes anyways. This seems like wishful thinking.
6. Add some magic to dask where it automatically logs warnings filters that were activated when a lazy function was added to a dask graph, and then restores them with executing the function. This sounds like the cleanest option, but it might have prohibitively large override. Possibly we could do this only for functions known to issue warnings, but if we were really careful about this it would be a very long list (e.g., including any numpy function that does division).

xref https://github.com/pydata/xarray/issues/1652

**Comment by WeatherGod** [comment]
I'd like to second this issue. I am having a hard time explaining to co-workers who aren't familiar with dask, but are reviewing my code why I have to add an extra set of warning filters in odd places that don't immediately appear to have anything to do with computation, such as places where data gets saved out.

Perhaps the dask object should record what the warnings filter is when the dask object was created and utilize it when `compute()` is called? (I know, easier said then done). Would be nice to also check numpy's errstate as well, but I don't know now it well enough to know if that can be introspected.

**Comment by mrocklin** [comment]
I have no strong feelings on this topic.  It looks like this issue needs two things:

1.  Discussion to determine which policy is best
2.  Someone to implement that policy

I encourage people to discuss this topic further to determine what policy is best.

**Comment by shoyer** [comment]
The best immediate solution might be simply redirecting all warnings to logging during evaluation of a dask graph, e.g., with [`logging.captureWarnings`](https://docs.python.org/3/library/logging.html#logging.captureWarnings). This at least makes it easy to customize how warnings are handled by dask all in one place, and to silence warnings issued via dask without silencing other warnings.

It would possibly also make sense to set a default NullHandler (per [these instructions](https://docs.python.org/3.7/howto/logging.html#configuring-logging-for-a-library)), which would ensure that the warnings are accessible but not won't end up in stderr unless explicitly requested. Or we could leave setting up the NullHandler to user code.

**Comment by gerritholl** [comment]
It's not enough to manage `warnings.catch_warnings`, one would also need to manage `np.errstate`, since either can be used to silence a `RuntimeWarning`.  Perhaps not in the example at the OP, but if I have

`np.array([1, 2, nan]) < 1.5`

then it's quite valid to use

```
with np.errstate(invalid="ignore"):
    np.array([1, 2, nan]) < 1.5
```

Unfortunately, this will have no effect if we replace the array by a dask array.

**Comment by gerritholl** [comment]
From a user point of view, ideally I would like something that mimics the numpy `errstate` context manager closely, so that a user could write:

```
with dask.errstate(invalid="ignore"):
    da = da.where(v <= 1.5)
```

I'm new to dask, but I imagine that dask would store a copy of `np.geterr()` with each task at submission time, then use this with `np.errstate()` when computing each task.  But that only works for cases that are covered by `np.errstate`, which `nanmean` is not; I suspect it may be more complicated to apply the same for warnings filters.

**Comment by jakirkham** [comment]
That sounds like a good approach. Are you interested in working on this, @gerritholl?

**Comment by gerritholl** [comment]
In principle, interested yes, but I can't allocate time to this right now (also because I would need to learn a lot more about dask first).

**Comment by mrocklin** [comment]
cc @jakirkham , you might find this interesting

**Comment by jakirkham** [comment]
Sure I've been tracking this for a little bit.

My understanding is we are looking for a solution like [this]( https://github.com/dask/dask/issues/3245#issuecomment-489646034 ). @shoyer, would you be able to confirm if this matches what you are looking for?

**Comment by shoyer** [comment]
@jakirkham Yes, a dask version of `np.errstate` would solve part of this problem.

More generally, we'd also like the ability to filter warnings more generally within a context manager, e.g., to handle cases like [this one](https://github.com/pydata/xarray/blob/66581084a89f75476b581ef74e5226eae2d62a84/xarray/core/duck_array_ops.py#L158) inside xarray:
```python
    with warnings.catch_warnings():
        warnings.filterwarnings('ignore', "In the future, 'NAT == x'")

        flag_array = (arr1 == arr2)
```

So you might want something like `dask.catch_warnings()` as well.

**Comment by jakirkham** [comment]
We can probably set some sort of config option to handle warnings. These also can be set with a context manager. `dask.catch_warnings()` could then be a user friendly way of setting that config option.

**Comment by shoyer** [comment]
I wonder if dask could safely copy the active value of warnings.filters
into the graph directly rather than having its own configuration option.

**Comment by WeatherGod** [comment]
+1 on this point. My original point isn't addressed here, and that is that
when I write up some generic function that expects numpy-like arrays, I
have already been setting the errstate for numpy. Then along comes a dask
array and all of the sudden, my errstate settings don't apply any more. I
don't want to have to import a dask package (and thereby have a dask
dependency where I didn't have one before), and have to set the errstate
twice on the off-chance that one of my inputs is actually a dask array
under the hood.

**Comment by shoyer** [comment]
We could definitely do this for NumPy with np.geterr. I don't know if
warnings.filters is a public API, though. Also there are potentially
performance implications to setting warnings filters in every task.

**Comment by astrofrog** [comment]
Just wanted to add a +1 to this feature request!

**Comment by jrdzha** [comment]
+1 here as well.. but am wondering if theres a way to suppress general warnings beyond those generated by np

**Comment by miroenev** [comment]
+1

**Comment by adriesse** [comment]
+1

**Comment by max-sixty** [comment]
Please do not leave +1s on issues. If you have something semantic to add, please feel free. Otherwise add a 👍 on the issue. Thanks.

**Comment by max-sixty** [comment]
(I added and deleted a comment because I thought this was the issue on the xarray repo, apologies)

**Comment by fingoldo** [comment]
Any solution? There are huge logs of such warnings in the ipython notebook...

**Comment by CptOrange16** [comment]
+1 on this issue, the the error logs are growing a lot because of the warning message

**Comment by vnoel** [comment]
When running my code I am getting lots of warnings like "dask/core.py:119: RuntimeWarning: invalid value encountered in true_divide". From the discussion above I understand there is currently no way to silence those warnings (np.seterr(invalid='ignore') does not work). Is that correct?

**Comment by j-bennet** [comment]
@bertcoerver Since you're defining `cve2` yourself, you can catch warnings before `np.nanmean` is called, just like it's done here:

https://github.com/dask/dask/pull/9916/files#diff-b60d0c0c32d24414e50ffbedb190c9e227b6b7db7e4727d1121b2e6bde68073fR645-R649

**Comment by bertcoerver** [comment]
Thanks @j-bennet but I tried that and doesnt work because Numba doesnt like it.

**Comment by vnoel** [comment]
I am sorry if I am a pain, but this issue is making my Jupyter notebooks really messy. The useful output is getting hard to find among the warnings. It's getting hard to work this way.
![Capture d’écran 2023-04-14 à 13 23 55](https://user-images.githubusercontent.com/731499/232031672-fc8eda54-420e-41f7-b5ca-9cfd930130e1.png)

**Comment by j-bennet** [comment]
@vnoel Which function are you getting those from?

**Comment by vnoel** [comment]
@j-bennet I am reading a lot of netCDF files with open_mfdataset() into a single dataset, and then dividing one of the dataset variables by another variable of the dataset.

**Comment by j-bennet** [comment]
@vnoel 

> Is there a way to request warning filtering once at the beginning of a notebook?

It's possible to ignore warnings globally in a notebook, by setting the PYTHONWARNINGS env var:

```python
%env PYTHONWARNINGS=ignore
```

**Comment by lucascolley** [comment]
Also bumping into this in SciPy!

> Add some magic to dask where it automatically logs warnings filters that were activated when a lazy function was added to a dask graph, and then restores them with executing the function. This sounds like the cleanest option, but it might have prohibitively large overhead. Possibly we could do this only for functions known to issue warnings, but if we were really careful about this it would be a very long list (e.g., including any numpy function that does division).

This would be great to see. The alternative solution for us looks like injecting `filterwarnings` conditionally on whether we are testing Dask arrays vs NumPy, but it would be a lot cleaner if Dask was considerate of `np.errstate` contexts itself.
